File: Data Structure/2. Linked-List/Singly/singlyLinkedList.py
# a singly linked list example

# Node class
class Node:
  # constructor
  def __init__(self, data):
    self.data = data
    self.next = None

# linked list class
class LinkedList(object):
  # constructor
  def __init__(self):
    self.head = None
  
  # check if the list is empty
  def is_empty(self):
    if self.head == None:
      return True
    return False

  # output the empty status
  def linked_list_status(self):
    if self.is_empty():
      print("The list is empty")
    else:
      print("The list is not empty")
    print # new line for better looking output

  # insert node to the front
  def insert_front(self, value):
    temp = Node(value) # create a new node
    temp.next = self.head # let head be the new node's next node
    self.head = temp # let temp be the new head

  # insert node to the back
  def insert_back(self, value):
    temp = Node(value) # create a new node

    # first check if the list is empty
    if self.is_empty():
      self.head = temp # let temp be the head because list is empty
      return
    
    # the list is not empty, iterate to the back
    currentNode = self.head
    while currentNode.next != None:
      currentNode = currentNode.next # go to next node until None
    
    # let the tail's next node be the new node
    currentNode.next = temp # insert to the back

  # insert after a specified value
  def insert_after_value(self, value, specified):
    # value = the new value
    # specified = the old specifed value node
    currentNode = self.head # let current be the head

    # iterate to find the location of specified value
    while currentNode != None and currentNode.data != specified:
      currentNode = currentNode.next # go to next node
    
    # if specified value not found
    if currentNode == None:
      print("Specified Value is not in the list!")
    else: # specified value is in the list
      temp = Node(value) # create a new node
      temp.next = currentNode.next # let new node points to specified's next node
      currentNode.next = temp # let specified node points to the new node

  # delete the front of the list
  def delete_front(self):
    # if list is empty
    if self.is_empty():
      print("List is empty!")
    else:
      temp = self.head
      self.head = self.head.next # let new head be the next node
      temp.next = None # let the old head points to nothing

  # delete the back of the list
  def delete_back(self):
    # if list is empty
    if self.is_empty():
      print("List is empty!")
      return

    # if only one node in the list
    if self.head.next == None:
      self.head = None
      return

    # if more than one exist
    currentNode = self.head
    while currentNode.next.next != None:
      currentNode = currentNode.next # go the location before tail
    currentNode.next = None

  # delete a specified value
  def delete_specified(self, specified):
    # if list is empty
    if self.is_empty():
      print("List is empty!")
      return
    
    # not empty, find it and delete it if exist
    # if head is the wanted node
    if self.head.data == specified:
      temp = self.head # record the head location
      self.head = self.head.next # let new head be the next node
      temp.next = None # let old head points to nothing
      return

    # if head is not the node and list length > 1
    currentNode = self.head
    previousNode = None
    while currentNode != None and currentNode.data != specified:
      previousNode = currentNode # record previous location
      currentNode = currentNode.next # go to next node until end or found location

    # if not found
    if currentNode == None:
      print("Specified value is not in the list!")
    else:
      previousNode.next = currentNode.next # let previousNode points to specifed's next
      currentNode.next = None # let specified node points to nothing

  # search a specified value
  def search(self, value):
    currentNode = self.head
    while currentNode != None and currentNode.data != value:
      currentNode = currentNode.next
    
    # if not found
    if currentNode == None:
      print("Search value is not in the list!");
    else:
      print("Search Value Founded!")
  
  # print the linked list values
  def print_list(self):
    # if the list is empty
    if self.is_empty():
      print("List is empty!")
      return

    # not empty, iterate and print it
    currentNode = self.head # let currentNode be the head node
    while currentNode.next != None:
      print(str(currentNode.data) + " ->"),
      currentNode = currentNode.next # go to next node until None
    print(currentNode.data) # print last node

# ============================
# testing the LinkedList class
# ============================

# create a empty linked list object
linked_list = LinkedList()
print("- Linked List Created")
linked_list.linked_list_status()

# insert elements to the front of the list
for i in range(0, 6):
  linked_list.insert_front(i)
print("- Elements are inserted to the front!")
linked_list.print_list()
linked_list.linked_list_status()

# insert elements to the back of the list
for i in range(-1, -6, -1):
  linked_list.insert_back(i)
print("- Elements are inserted to the back!")
linked_list.print_list()
linked_list.linked_list_status()

# insert element after a specified value
print("- Try insert -7 after value -6!")
linked_list.insert_after_value(-7, -6)
print("- Try insert -6 after value -5!")
linked_list.insert_after_value(-6, -5)
linked_list.print_list()
linked_list.linked_list_status()

# delete the front of the list
linked_list.delete_front()
print("- The front of the list is deleted!")
linked_list.print_list()
linked_list.linked_list_status()

# delete the back of the list
linked_list.delete_back()
print("- The back of the list is deleted!")
linked_list.print_list()
linked_list.linked_list_status()

# delete a specified value of the list
linked_list.delete_specified(-5)
print("- Value -5 of the list is deleted!")
linked_list.print_list()
linked_list.linked_list_status()

# search a specified value of the list
print("- Try search -5!")
linked_list.search(-5)
print("- Try search -4!")
linked_list.search(-4)
linked_list.print_list()
linked_list.linked_list_status()

File: Data Structures/2. Linked-List/Singly/singlyLinkedList.cpp
#include <iostream>
using namespace std;

struct Node {
  int data;
  Node* next;
  // node constructor
  Node(int value): data(value), next(NULL) {}
} *head = NULL;

// check if Linked list is empty
bool isEmpty() { return (head == NULL); }

// insert value to the front
void insertFront(int value) {
  Node* temp = new Node(value); // create a temporary node
  temp -> next = head; // node points to the front
  head = temp; // let new node be the head
}

// insert value to the back
void insertBack(int value) {
  Node* temp = new Node(value); // create a temporary node
  
  // first check if list is empty
  if (isEmpty()) {
    head = temp; // let temp be the first node
    return;
  }

  // not empty list, insert back
  Node* current = head;
  while (current -> next != NULL) // go the last node
    current = current -> next;
  current -> next = temp; // let temp be the last node
}

// insert value after a desired value
void insertAfterValue(int value, int desiredValue) {
  // iterate to desired value location (if exists)
  Node* current = head; 
  while (current != NULL && current -> data != desiredValue)
    current = current -> next; // go to next available node

  // if the desired value does not exist
  if (current == NULL) {
    cout << "Desired value " << desiredValue << " does not exist!" << endl;
  }
  // the desired value node exist, insert after it
  else {
    Node* temp = new Node(value);
    temp -> next = current -> next; // temp points to desired next location
    current -> next = temp; // let desired value node points to temp
    // original: Desired -> nextNode
    // now: Desired -> temp -> nextNode
  }
}

// delete the head of a linked list
void deleteHead() {
  if (isEmpty()) {
    cout << "List is empty! No need for deletion!" << endl;
  }
  // else the list is not empty, delete head
  else {
    Node* temp = head;
    head = head -> next; // let new head be the next node
    temp -> next = NULL; // let original head points to NULL
    delete temp; // delete original head in memory
  } 
}

// delete the tail of a linked list
void deletTail() {
  if (isEmpty())
    cout << "List is empty! No need for deletion!" << endl;

  // if head is the only node, delete head
  if (head -> next == NULL) {
    delete head;
    head = NULL;
    return;
  }
  // if the list has > 1 nodes, delete tail
  Node* temp = head;
  // let temp go the location before tail
  while (temp -> next -> next != NULL)
    temp = temp -> next;
  delete temp -> next; // delete tail
  temp -> next = NULL; // set temp to tail
}

// delete the whole list
void deleteList() {
  while (!isEmpty())
  	deleteHead();
}

// delete desired value in linked list
void deleteDesiredValue(int desiredValue) {
  // first check if empty
  if (isEmpty()) {
    cout << "List is empty! Can't delete nothing!" << endl;
    return;
  }

  // create temp and prev node for iterations
  Node* temp = head;
  Node *prev = NULL;

  // if the head is the desired value
  if (head -> data == desiredValue) {
    head = head -> next; // let the next node be new head
    temp -> next = NULL; // let original head points to NULL
    delete temp; // delete original head
    return;
  }

  // iterate the linked list to find the desiredValue
  while (temp != NULL && temp -> data != desiredValue) {
    prev = temp; // set prev to the temp node
    temp = temp -> next; // let temp be the next node
  }

  // desired value does not exist
  if (temp == NULL) {
    cout << "Desired value does not exist in linked list" << endl;
  }
  // desire value exist, delete it
  else {
    prev -> next = temp -> next; // let prev node points to temp's next node
    temp -> next = NULL; // let desired node points to NULL
    delete temp; // delete the desired node
  }
}

// search the desired value
void search(int desiredValue) {
  Node* temp = head;
  // iterate the list to find the desired value
  while (temp != NULL && temp -> data != desiredValue)
    temp = temp -> next;
  
  // if the desired value does not exist
  if (temp == NULL)
    cout << "Desired Value Not Found!" << endl;
  else
    cout << "Desired Value Found in the list" << endl;
}

// find the linked list length
void listLength() {
  int length = 0;
  Node* temp = head;
  while (temp != NULL) {
    length += 1;
    temp = temp -> next;
  }
  cout << "List Length is: " << length << endl;
}

// find the linked list length recursively
int LengthRecursiveHelper(Node* current) {
  if (current == NULL) return 0;
  return (1 + LengthRecursiveHelper(current -> next));
}

int listLengthRecursive() {
  int length = LengthRecursiveHelper(head);
  cout << "List Length is " << length << endl;
}

// output list
void outputList() {
  // edge case: list is empty
  if (isEmpty()) {
    cout << "List is empty!" << endl;
    return;
  }

  Node* current = head;
  while (current != NULL) {
    cout << current -> data; // output current node's value
    current = current -> next; // move to next node
    if (current != NULL) cout << " -> ";
  }
  cout << endl;
}

int main() {
  // insert in the front
  cout << "\nDemonstrate insertFront() function\n";
  for (int i = 1; i < 5; i++) 
    insertFront(i);
  outputList(); // print the linked list

  // insert at the end
  cout << "\nDemonstrate insertBack() function\n";
  for (int i = 2; i < 5; i++) 
    insertBack(i);
  outputList(); // print the linked list

  // insert value after 1
  cout << "\nDemonstrate insertAfterValue() function\n";
  insertAfterValue(1, 1);
  outputList(); // print the linked list

  // delete tail
  cout << "\nDemonstrate deleteTail() function\n";
  deletTail();
  outputList(); // print the linked list

  // delete head
  cout << "\nDemonstrate deleteHead() function\n";
  deleteHead();
  outputList(); // print the linked list

  // delete a specified value
  cout << "\nDemonstrate deleteDesiredValue() function\n";
  deleteDesiredValue(1);
  outputList(); // print the linked list

  // search a specified value
  cout << "\nDemonstrate search() function\n";
  search(1);

  // length of a linkedin list
  cout << "\nFinding length iteratively: length = ";
  listLength(); 
  cout << "Finding length recursively: length = ";
  listLengthRecursive(); 

  // delete the list
  cout << "\nDemonstrate deleteList() function\n";
  deleteList();
  outputList(); // print the linked list
}